internal/app/export/service.go

```diff
@@ -4,62 +4,248 @@
+	connected
+)
+
 // New create new instance of export service
-func New(mqtt mqtt.Client, c config.Config, logger log.Logger) Service {
-	routes := make([]routes.Route, 0)
+func New(c config.Config, cache messages.Cache, l logger.Logger) Service {
+	routes := make(map[string]routes.Route)
+	id := fmt.Sprintf("export-%s", c.MQTT.Username)
+
 	e := exporter{
-		Mqtt:   mqtt,
-		Logger: logger,
-		Cfg:    c,
-		Routes: routes,
+		ID:        id,
+		Logger:    l,
+		Cfg:       c,
+		Consumers: routes,
+		Cache:     cache,
+		connected: make(chan bool, 1),
 	}
-
+	client, err := e.mqttConnect(c, l)
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	e.Mqtt = client
 	return &e
 }
 
-// Start method starts consuming messages received from NATS
-// and makes routes according to the configuration file.
-// Routes export messages to mqtt.
-func (e *exporter) Start(queue string, nc *nats.Conn) {
+// Start method loads route configuration
+func (e *exporter) Start(queue string) {
 	var route routes.Route
 	for _, r := range e.Cfg.Routes {
 		switch r.Type {
 		case "mfx":
-			route = mfx.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
+			route = mfx.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic)
 		default:
-			route = routes.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
+			route = routes.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic)
 		}
-		e.Routes = append(e.Routes, route)
-		e.Subscribe(route, queue, nc)
+
+		if e.validateTopic(route.NatsTopic()) {
+			e.Consumers[route.NatsTopic()] = route
+			g, err := e.Cache.GroupCreate(r.NatsTopic, exportGroup)
+			if err != nil {
+				e.Logger.Error(fmt.Sprintf("Failed to create stream group %s", err.Error()))
+			}
+			e.Logger.Info(fmt.Sprintf("Stream group %s created %s", route.NatsTopic(), g))
+		}
+
 	}
+	e.startRepublish()
 }
 
-func (e *exporter) Subscribe(r routes.Route, queue string, nc *nats.Conn) {
-	_, err := nc.QueueSubscribe(r.NatsTopic(), fmt.Sprintf("%s-%s", queue, r.NatsTopic()), r.Consume)
+func (e *exporter) Consume(msg *nats.Msg) {
+	if _, ok := e.Consumers[msg.Subject]; !ok {
+		e.Logger.Info(fmt.Sprintf("no configuration for nats topic %s", msg.Subject))
+		return
+	}
+
+	route := e.Consumers[msg.Subject]
+	payload, err := route.Consume(msg)
 	if err != nil {
-		e.Logger.Error(fmt.Sprintf("Failed to subscribe for NATS/MQTT %s/%s", r.NatsTopic(), r.MqttTopic()))
+		e.Logger.Error(fmt.Sprintf("Failed to consume msg %s", err.Error()))
+	}
+
+	if err = e.publish(route.MqttTopic(), payload); err != nil {
+		_, err := e.Cache.Add(msg.Subject, route.MqttTopic(), payload)
+		if err != nil {
+			e.Logger.Error(fmt.Sprintf("Failed to add to redis stream `%s`", msg.Subject))
+		}
+		return
+	}
+}
+
+func (e *exporter) startRepublish() {
+	// initial connection established on start up
+	<-e.connected
+	e.Logger.Info("Starting republish, waiting for stream data")
+	for _, route := range e.Cfg.Routes {
+		streams := []string{route.NatsTopic, ">"}
+		go func() {
```

Comment:
Defining function here does not help readability IMHO

Reply:
i think this is absolutely go idiomatic , spawning a routine over a small piece of code

Reply:
This will not be accepted.

Reply:
@dusanb94 @nmarcetic @anovakovic01 @manuio please review this

Reply:
For me, this is OK, especially because of loops.

Reply:
I agree with @drasko. While this kind of anonymous function is OK, can you consider having this function as an unexported method just for the sake of readability? Something like:
```go
func(e *exporter) republish() {
        streams := []string{route.NatsTopic, ">"}
	for {
		msgs, err := e.readMessages(streams)
		if err != nil {
			continue
		}
		e.Logger.Info(fmt.Sprintf("Waiting for connection to %s", e.Cfg.MQTT.Host))
		for {
			// Wait for connection
			if e.IsConnected() || <-e.connected {
				for _, m := range msgs {
					if err := e.publish(m.Topic, []byte(m.Payload)); err != nil {
						e.Logger.Error("Failed to republish message")
					}
				}
				break
			}
		}
	}
}
```